drivers/wdf/wdfldr/class.c

```diff
@@ -0,0 +1,895 @@
+
+	if (ClassModule->Service.Buffer != NULL)
+	{
+		ExFreePoolWithTag(ClassModule->Service.Buffer, 0);
+		ClassModule->Service.Length = 0;
+		ClassModule->Service.Buffer = 0;
+	}
+
+	if (ClassModule->ImageName.Buffer != NULL)
+	{
+		ExFreePoolWithTag(ClassModule->ImageName.Buffer, 0);
+		ClassModule->ImageName.Length = 0;
+		ClassModule->ImageName.Buffer = 0;
+	}
+
+	ExDeleteResourceLite(&ClassModule->ClientsListLock);
+	ExFreePoolWithTag(ClassModule, 0);
+}
+
+
+PCLASS_MODULE
+NTAPI
+ClassCreate(
+	IN PWDF_CLASS_LIBRARY_INFO ClassLibInfo,
+	IN PLIBRARY_MODULE LibModule,
+	IN PUNICODE_STRING ServiceName
+)
+{
+	PCLASS_MODULE result;
+	NTSTATUS status;
+
+	result = ExAllocatePoolWithTag(NonPagedPool, sizeof(CLASS_MODULE), WDFLDR_TAG);
+
+	if (result != NULL)
+	{
+		goto exit;
+	}
+
+	memset(result, 0, sizeof(CLASS_MODULE));
+	if (ClassLibInfo != NULL)
+	{
+		result->ImplicitlyLoaded = TRUE;
+		result->ClassLibraryInfo = ClassLibInfo;
+		result->Version.Major = ClassLibInfo->Version.Major;
+		result->Version.Minor = ClassLibInfo->Version.Minor;
+		result->Version.Build = ClassLibInfo->Version.Build;
+	}
+
+	result->ClassRefCount = 1;
+	result->Library = LibModule;
+	ExInitializeResourceLite(&result->ClientsListLock);
+	InitializeListHead(&result->ClientsListHead);
+	InitializeListHead(&result->LibraryLinkage);
+	result->Service.Buffer = ExAllocatePoolWithTag(PagedPool, ServiceName->MaximumLength, WDFLDR_TAG);
+
+	if (result->Service.Buffer == NULL)
+	{
+		goto clean;
+	}
+
+	result->Service.MaximumLength = ServiceName->MaximumLength;
+	memset(result->Service.Buffer, 0, result->Service.MaximumLength);
+	RtlCopyUnicodeString(&result->Service, ServiceName);
+	status = GetImageName(ServiceName, 0x694C7846u, &result->ImageName);
+
+	if (!NT_SUCCESS(status))
+	{
+		goto clean;
+	}
+
+	result->IsBootDriver = ServiceCheckBootStart(&result->Service);
+	status = GetImageBase((PCUNICODE_STRING)& result->ImageName, &result->ImageAddress, &result->ImageSize);
+
+	if (NT_SUCCESS(status))
+	{
+		return result;
+	}
+	else if (WdfLdrDiags)
+	{
+			DbgPrint("WdfLdr: ClassCreate - ");
+			DbgPrint("WdfLdr: ClassCreate: GetImageBase failed\n");
+	}
+
+clean:
+	ClassCleanupAndFree(result);
+exit:
+	return NULL;
+}
+
+
+PCLASS_CLIENT_MODULE
+NTAPI
+ClassClientCreate()
+{
+	PCLASS_CLIENT_MODULE result;
+
+	result = ExAllocatePoolWithTag(NonPagedPool, sizeof(CLASS_CLIENT_MODULE), WDFLDR_TAG);
+
+	if (result != NULL)
+	{
+		memset(result, 0, sizeof(CLASS_CLIENT_MODULE));
+		InitializeListHead(&result->ClassLinkage);
+		InitializeListHead(&result->ClientLinkage);
+	}
+
+	return result;
+}
+
+
+PCLIENT_MODULE
+NTAPI
+LibraryFindClientLocked(
+	IN PLIBRARY_MODULE LibModule,
+	IN PWDF_BIND_INFO BindInfo
+)
+{
+	PCLIENT_MODULE result;
+
+	result = (PCLIENT_MODULE)LibModule->ClientsListHead.Flink;
+
+	for (; result != (PCLIENT_MODULE)&LibModule->ClientsListHead;
+		result = (PCLIENT_MODULE)result->LibListEntry.Flink)
+	{
+
+		if (result->Info == BindInfo)
+			return result;
+	}
```

Comment:
I don't really like this list enumeration. Can you do it more properly without casting and using CONTAINING_RECORD macro?